File: 2015/day_5/d5_p1_solution.py
with open(r'./d5_input.txt', 'r') as f:
    inpt = f.read()
    
lines = inpt.splitlines()

def is_vowel(c) -> bool:
    return c.lower() in 'aeiou'
    
def three_vowels(string: str) -> bool:
    vowels = 0
    for c in string:
        if is_vowel(c):
            vowels += 1
            
    if vowels >= 3:
        return True
    else:
        return False
    
def has_double(string: str) -> bool:
    for i, c in enumerate(string[:-1]):
        if c == string[i+1]:
            return True
    return False
    
def no_nasties(string: str) -> bool:
    for i, c in enumerate(string[:-1]):
        if (c + string[i+1]).lower() in ['ab', 'cd', 'pq', 'xy']:
            return False
    return True
    
def check_string(string: str) -> bool:

    has_three = three_vowels(string)
    has_doub = has_double(string)
    no_foul = no_nasties(string)
    
    if has_three and has_doub and no_foul:
        return True
    else:
        return False

nice_strings = 0
for line in lines:
    if check_string(line):
        nice_strings += 1
        
print(f"There are {nice_strings} nice strings.")
    

File: 2015/day_5/d5_p1_solution.rs
use std::fs;

const VOWELS: &str = "aeiou";
const NAUGHTIES: [[char; 2]; 4] = [['a', 'b'],
                                   ['c', 'd'],
                                   ['p', 'q'],
                                   ['x', 'y']];

fn is_vowel(c: char) -> bool {

    return VOWELS.contains(c)
    
}

fn three_vowels(s: &str) -> bool {
   
    let mut num_vowels = 0;
    for c in s.to_lowercase().chars() {
        if is_vowel(c) {num_vowels += 1}
    }
    num_vowels >= 3
   
}

fn has_double(s: &str) -> bool {
    s.chars()
     .collect::<Vec<char>>()
     .windows(2)
     .any(|w| w[0] == w[1])
}

fn has_naughty(s: &str) -> bool {
    s.to_lowercase()
     .chars()
     .collect::<Vec<char>>()
     .windows(2)
     .any(|i| NAUGHTIES.contains(&[i[0], i[1]]))
}

fn main() {

    let mut nice_strings = 0;
    fs::read_to_string("./d5_input.txt")
        .expect("Should have been able to read line.")
        .as_str()
        .lines()
        .for_each(|i| {
            if three_vowels(&i) & has_double(&i) & !has_naughty(&i) { nice_strings += 1}
        });
    println!("There are {} nice strings.", nice_strings);
}
